Allow RPC framework to use rank in addition to WorkerInfo and name.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46221 Allow RPC framework to use rank in addition to WorkerInfo and name.**

The RPC framework only allowed sending RPCs based on provided
WorkerInfo or name. When using RPC with DDP, sometimes it might just be easier
to refer to everything in terms of ranks since DDP doesn't support names yet.

As a result, support a `to` parameter in the RPC APIs which allow for
specifying a rank as well would be helpful.

Differential Revision: [D24264989](https://our.internmc.facebook.com/intern/diff/D24264989/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24264989/)!

pritamdamania committed:
Allow RPC framework to use rank in addition to WorkerInfo and name.

The RPC framework only allowed sending RPCs based on provided
WorkerInfo or name. When using RPC with DDP, sometimes it might just be easier
to refer to everything in terms of ranks since DDP doesn't support names yet.

As a result, support a `to` parameter in the RPC APIs which allow for
specifying a rank as well would be helpful.

Differential Revision: [D24264989](https://our.internmc.facebook.com/intern/diff/D24264989/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24264989/)!

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -417,7 +417,7 @@ def remote(to, func, args=None, kwargs=None, timeout=UNSET_RPC_TIMEOUT):
 
     Arguments:
-        to (str or WorkerInfo): id or name of the destination worker.
+        to (str or WorkerInfo or int): name/rank/WorkerInfo of the destination worker.
```
double quote code, only for the one after colon

```
``WorkerInfo``
```

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -672,7 +672,7 @@ def rpc_sync(to, func, args=None, kwargs=None, timeout=UNSET_RPC_TIMEOUT):
 
     Arguments:
-        to (str or WorkerInfo): id or name of the destination worker.
+        to (str or WorkerInfo or int): name/rank/WorkerInfo of the destination worker.
```
ditto

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -751,7 +751,7 @@ def rpc_async(to, func, args=None, kwargs=None, timeout=UNSET_RPC_TIMEOUT):
 
     Arguments:
-        to (str or WorkerInfo): id or name of the destination worker.
+        to (str or WorkerInfo or int): name/rank/WorkerInfo of the destination worker.
```
ditto

mrshenli reviewed (approved):
LGTM!

pritamdamania committed:
Update on "Allow RPC framework to use rank in addition to WorkerInfo and name."


The RPC framework only allowed sending RPCs based on provided
WorkerInfo or name. When using RPC with DDP, sometimes it might just be easier
to refer to everything in terms of ranks since DDP doesn't support names yet.

As a result, support a `to` parameter in the RPC APIs which allow for
specifying a rank as well would be helpful.

Differential Revision: [D24264989](https://our.internmc.facebook.com/intern/diff/D24264989/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24264989/)!

[ghstack-poisoned]
